source/backend/api/Areas/Documents/DocumentController.cs

```diff
@@ -90,16 +90,36 @@ public IActionResult GetDocumentTypes()
         }
 
         /// <summary>
-        /// Downloads the file for the corresponding file and document id.
+        /// Downloads the file for the corresponding file and document id wrapped and encoded base64.
         /// </summary>
-        [HttpGet("storage/{mayanDocumentId}/files/{mayanFileId}/download")]
+        [HttpGet("storage/{mayanDocumentId}/files/{mayanFileId}/download-wrapped")]
         [HasPermission(Permissions.DocumentView)]
         [ProducesResponseType(typeof(ExternalResult<FileDownload>), 200)]
         [SwaggerOperation(Tags = new[] { "storage-documents" })]
+        public async Task<IActionResult> DownloadWrappedFile(long mayanDocumentId, long mayanFileId)
+        {
+            var result = await _documentService.DownloadFileAsync(mayanDocumentId, mayanFileId);
+            return new JsonResult(result.Payload);
+        }
+
+        /// <summary>
+        /// Downloads the file for the corresponding file and document id and returns a stream.
+        /// </summary>
+        [HttpGet("storage/{mayanDocumentId}/files/{mayanFileId}/download")]
+        [HasPermission(Permissions.DocumentView)]
+        [ProducesResponseType(typeof(FileContentResult), 200)]
+        [SwaggerOperation(Tags = new[] { "storage-documents" })]
         public async Task<IActionResult> DownloadFile(long mayanDocumentId, long mayanFileId)
         {
             var result = await _documentService.DownloadFileAsync(mayanDocumentId, mayanFileId);
-            return new FileStreamResult(result.Payload.FilePayload, result.Payload.Mimetype) { FileDownloadName = result.Payload.FileName };
+
+            if (result?.Payload == null)
+            {
+                return new NotFoundResult();
+            }
+
+            byte[] fileBytes = System.Convert.FromBase64String(result.Payload.FilePayload);
```

Comment:
This will force the entire stream to be stored in memory before it can be returned. It also forces the json parser to parse the entire object instead of parsing a stream. In the short term this will probably cause file downloads of >10mb to fail.

Can we come up with a way of using file streams instead? to me, a file stream has significantly lower impact of the memory usage of the server.

Reply:
Yes we can. We could keep the stream from the repository and have the service be smarter to if it needs to be serialized to base64.